lightning/src/routing/router.rs

```diff
@@ -78,10 +83,134 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
+					.map(|info| &info.channels[..])
+					.and_then(|channels| (channels.len() >= MIN_PEER_CHANNELS).then(|| channels))
+					.map(|channels| (details, counterparty_node_id, channels))
+			)
+			// Pair counterparties with their other channels
+			.flat_map(|(details, counterparty_node_id, counterparty_channels)|
+				counterparty_channels
+					.iter()
+					.filter_map(|scid| network_graph.channels().get_key_value(scid))
+					.filter_map(move |(scid, info)| info
+						.as_directed_to(&counterparty_node_id)
+						.map(|(info, source)| (source, *scid, info))
+					)
+					.filter(|(source, _, _)| **source != recipient_node_id)
+					.filter(|(source, _, _)| network_graph
+						.node(source)
+						.and_then(|info| info.announcement_info.as_ref())
+						.map(|info| info.features.supports_route_blinding())
+						.unwrap_or(false)
+					)
+					.filter(|(_, _, info)| amount_msats >= info.direction().htlc_minimum_msat)
+					.filter(|(_, _, info)| amount_msats <= info.direction().htlc_maximum_msat)
+					.map(move |(source, scid, info)| (source, scid, info, details))
+			)
+			// Construct blinded paths where the counterparty's counterparty is the introduction
+			// node:
+			//
+			// source --- info ---> counterparty --- details ---> recipient
+			.map(|(introduction_node_id, scid, info, details)| {
+				let counterparty_forward_node = {
+					let short_channel_id = details.get_inbound_payment_scid().unwrap();
+					let payment_relay: PaymentRelay =
+						details.counterparty.forwarding_info.clone().unwrap().into();
+					let payment_constraints = PaymentConstraints {
+						max_cltv_expiry: payment_relay.cltv_expiry_delta as u32
+							+ tlvs.payment_constraints.max_cltv_expiry,
+						htlc_minimum_msat: details.inbound_htlc_minimum_msat.unwrap_or(0),
+					};
+					ForwardNode {
+						tlvs: ForwardTlvs {
+							short_channel_id,
+							payment_relay,
+							payment_constraints,
+							features: BlindedHopFeatures::empty(),
+						},
+						node_id: details.counterparty.node_id,
+						htlc_maximum_msat: details.inbound_htlc_maximum_msat.unwrap_or(0),
+					}
+				};
+				let introduction_forward_node = {
+					let htlc_minimum_msat = info.direction().htlc_minimum_msat;
+					let htlc_maximum_msat = info.direction().htlc_maximum_msat;
+					let payment_relay: PaymentRelay = info.into();
+					let payment_constraints = PaymentConstraints {
+						max_cltv_expiry: payment_relay.cltv_expiry_delta as u32
+							+ counterparty_forward_node.tlvs.payment_constraints.max_cltv_expiry,
+						htlc_minimum_msat,
+					};
+					ForwardNode {
+						tlvs: ForwardTlvs {
+							short_channel_id: scid,
+							payment_relay,
+							payment_constraints,
+							features: BlindedHopFeatures::empty(),
+						},
+						node_id: introduction_node_id.as_pubkey().unwrap(),
+						htlc_maximum_msat,
+					}
+				};
+				BlindedPath::new_for_payment(
+					&[introduction_forward_node, counterparty_forward_node], recipient,
```

Comment:
It seems potentially likely for users with some `LDKServer<>LDKClient` architecture that the counterparty will support route blinding, but none of their public counterparties do. I don't think we should fail to construct a path in that situation?

Reply:
> How does the architecture play a role here?

The architecture/situation outlined seems like a somewhat common LDK use case where 3-hop wouldn't work, and 1-hop also wouldn't work (due to the client likely being private). 

To me it would make more sense if we always tried for 2-hop in this method, and fall back to 1-hop if the recipient is a public node, otherwise error. Then we wouldn't need to have 1-hop fallback logic in channelmanager as well. I may be missing something here, though.

Reply:
> > How does the architecture play a role here?
> 
> The architecture/situation outlined seems like a somewhat common LDK use case where 3-hop wouldn't work, and 1-hop also wouldn't work (due to the client likely being private).

Still a little confused about this. I thought you were talking about splitting an LDK node. But are you saying `LDKServer` is an LSP node and `LSPClient` is a separate node with a private channel with the LSP?

> To me it would make more sense if we always tried for 2-hop in this method, and fall back to 1-hop if the recipient is a public node, otherwise error. Then we wouldn't need to have 1-hop fallback logic in channelmanager as well. I may be missing something here, though.

When would we try 3-hops? Are you thinking something like this?

```rust
let paths = if has_public_channels {
    create_three_hop_paths()
} else {
    create_two_hop_paths()
};

if !paths.is_empty() {
    paths
} else if has_public_channels {
    create_one_hop_path()
} else {
    Err(())
}
```

Reply:
See https://github.com/lightningdevkit/rust-lightning/pull/2781#issuecomment-1855086265